streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.FLATMAP;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.GROUP_BY;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.MAP;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.SELECT_KEY;
+import static org.apache.kafka.streams.kstream.internals.TopologyNodeType.TRANSFORM;
+
+/**
+ * A Node in the Streams DAG used to contain all information needed to construct this
+ * portion of the Topology.  The class holds information until the InternalTopologyBuilder builds
+ * the actual topology.  This is the base case class and other more specific cases required
+ * more detailed information will extend this class
+ */
+public class StreamsGraphNode {
```

Comment:
Currently we have the following extensions of this node:

1. KTableJoinGN
2. SourceSinkGN
3. StreamStreamJoinGN

And these classes seem to be used for some specialized optimizations, but won't that restrict the flexibility of the optimizer itself? I suggest we make general extensions of the GN that optimizers can understand and manipulate for. Please see my other comments regarding StreamGN and ProcessDetails.

Reply:
In this case, I don't think so.  

The `SourceSinkNode` is used represent a repartition graph node and the name will be updated accordingly.

In the case of joins, IMHO we need specific classes to represent these actions, as there is too much information critical for performing these operations to make generalized nodes.